### ISSUE TITLE
RFC: Refactoring plan

### ISSUE THREAD

**Issue by Art4** [friendica/friendica#14535]
This is a WIP refactoring plan I came up for Friendica. Comments are very welcome.

## The current situation

At the moment we have the following situation in the `index.php` (simplified):

```php
// Dice is created and loaded with the rules
$dice = (new Dice())->addRules(include __DIR__ . '/static/dependencies.config.php');

// Dice is places inside the `DI` class.
\Friendica\DI::init($dice);

// App is created by calling DI::app()/Dice::create()
$a = \Friendica\DI::app();

// App runs a method to create the Frontend
$a->runFrontend(
	$dice->create(\Friendica\App\Router::class),
	$dice->create(\Friendica\Core\PConfig\Capability\IManagePersonalConfigValues::class),
	$dice->create(\Friendica\Security\Authentication::class),
	$dice->create(\Friendica\App\Page::class),
	$dice->create(\Friendica\Content\Nav::class),
	$dice->create(Friendica\Module\Special\HTTPException::class),
	new \Friendica\Util\HTTPInputData($_SERVER),
	$start_time,
	$_SERVER
);
```
This code has many down sides that leads to a lot of problems: Friendica uses Dice as a dependency injection container (DIC), what is generally great. But this DIC is placed inside the `Friendica\DI` class and then used as a service locator. By calling the static `DI` methods all over the place the DIC is used as a Singleton-Store instead of a dependeny injection container. This behavior is a wrong use of the DIC pattern. Also the "singleton-like" store makes it hard to mock dependencies in tests. Singletons in this project should be strongly considered as an anti-pattern.

## Refactoring plan

I've came up with this refactoring plan so far. This is a WIP and will be get more details the more I learn about the code base.

### 0. Stabilize source code and addons

Set a minimum bar for the code quality and check every code change. This should prevent us from regressions like #14566.

- [x] Reach PHPStan level 3 for friendica
  - [x] #14529
  - [x] #14559
  - [x] #14584
  - [x] #14830
- [x] Reach PHPStan level 3 for addon repository
  - [x] #14571
  - [x] #14581
  - [x] #14590
  - [x] https://git.friendi.ca/friendica/friendica-addons/pulls/1603
- [x] #14966
- [x] #14526
- [x] #14597
- [x] #14649 (will be done in multiple steps while refactoring)
- [x] #14694
- [x] Introduce Rector: #14989
- [ ] #14722

### 1. Extract most/all `App` methods into `AppHelper`

(The name `AppHelper` is just a quick idea in lack of imagination. It could also be named `LegacyApp`. Naming things is hard :blush:)

Extract all `App` methods, that are used somewhere in `src` folder into a new `AppHelper` class. The App class should be nearly empty and the AppHelper is just a store for the current code. It will be later refactored into other places like Models, Container, Services, etc.

- [x] #14561
- [x] #14542
- [x] Replace type hinting for `Friendica\App` with `Friendica\AppHelper` and calls for `DI::app()` with `DI::appHelper()`
  - [x] #14570 
  - [x] https://github.com/friendica/friendica-addons/pull/1373
  - [x] #14593

### 2. Remove `\Friendica\App` from Dice and `DI` class.

At the moment the `Dice` objects and `DIC` class are aware of the `App` class. This should be removed. Places where App methods are called have to be replaced with the `AppHelper` from step 1.

- [x] #14603
- [x] #14602
- [x] #14619

### 3. Create `App::fromDice()`

The index.php and the other entry points should be the only places that create the App object and call their methods. We are creating the `App` class in the index.php by injecting `Dice` (like `$app = App::fromDice($dice);`)

- [x] #14616
- [x] Refactor all entry points to use `App` class
  - [x] `index.php`: #14620
  - [x] `bin/auth_ejabberd.php`: #14623
  - [x] `bin/console.php`: #14625
  - [x] `bin/daemon.php`: #14626
  - [x] #14642
  - [x] `bin/jetstream.php`: #14637
  - [x] #14655
  - [x] `bin/worker.php`: #14628
  - [x] #14659
  - [ ] Move auth_ejabberd.php to Console class: #14920
  - [x] #14677

### 4. Replace Hook system with PSR-14 Event Dispatcher

The addons are to tight connected to the friendica code. Using an Event Dispatcher ([PSR-14](https://www.php-fig.org/psr/psr-14/)) we should replace the current Hook system to keep them separate.

- [x] #14640
- [x] #14799
- [x] #14853
- [ ] Replace all `Hook::callAll()` calls with EventDispatcher

### 5. Refactor Logger system

- [x] #14682
- [x] Replace `Friendica\Core\Logger` with constructor injection or `DI::logger`: #14688
- [x] Refactor all addons to provide a `LoggerInterface` implementation via `LoggerFactory`, https://git.friendi.ca/friendica/friendica-addons/pulls/1598
- [x] #14882 Rewrite `StreamLogger` and `SyslogLogger` to implement `LoggerFactory`

### 6. Refactor Addon system

- [x] #14770
- [x] #14795
- [x] #14909
- [x] Replace `Friendica\Core\Addon` class with `Friendica\Core\Addon\AddonHelper` in addons, see https://git.friendi.ca/friendica/friendica-addons/pulls/1607
- [x] #14913
- [x] #14937
- [ ] #14724

### 7. Encapsulate DICE in PSR-11 Container

We can now encapsulate DICE into a PSR-11 Container. The creation of the DICE object could be moved into a bootstrap file, that simply delivers a PSR-11 container

- [x] Create factory method `Friendica\App::fromContainer()`, done in #14659.
- [x] Move DICE instantiation into Container class, see #14659.
- [ ] Remove dependencies for Dice inside all classes (remove all appearances of `Dice::SELF` in `static/dependencies.config.php`)
  - [x] #14635
  - [ ] Remove the need for `\Friendica\Core\Hooks\Model\DiceInstanceManager`

### 8. Refactor Modules into Controller 

The `BaseModul` does to much and has some unnecessary dependencies. The implementation of `ICanHandleRequests` was a good start but has some downsides. I would recommend a controller system that allows us to create Controller classes with only the dependencies we need and also allow a controller to react on multiple similar request to avoid many "nano-controllers" only for a redirect or to share the code for similar request (e.g. creating and updating an item).

All controllers will be in the `Friendica\Controller` namespace:

- Friendica\Controller\Controller: An interface every controller has to be implement
- Friendica\Controller\AbstractController: An abstract class that contains some useful methods like `createRedirectResponse()` or `renderTemplate()`
- Friendica\Controller\Frontend\: Namespace for all frontend requests
- Friendica\Controller\API\: namespace for API request
- ...

### 9. Deprecate the `DI` class

Deprecating things and show the alternative is important. Instead of changing all things at ones in one hundreds of files is allows us to split this huge task into many many steps. The deprecated methods will stay as an legacy layer until every call of the deprecated methods are replaced.

- [ ] Deprecate every getter method in `Friendica\DI`
- [ ] Deprecate `Friendica\DI`

### More Refactoring

This step will be the most time consumption but can also be made step by step in multiple PRs without breaking anything.

Replace every `DI::...()` call with dependency injection using Dice. At the end the App class should be the only place aware of the DIC.

An IDE can inform us about the usage of deprecated methods and allows to replace it with the alternative. No new code should be created using deprecated methods/classes. Instead dependency injection should be used.

While refactoring more questions will arise how to achieve IoC (Inversion of Control) in a specific class or how to write tests and mock dependencies. This will make the [code smell](https://refactoring.guru/refactoring/smells) obvious. The answers to this questions will be multiple, like splitting classes into separate services/models/controller or using other [design patterns](https://refactoring.guru/design-patterns).

- [x] #14622
- [ ] Fix code style
- [ ] Refactor the functions from `/mod` folder into `Friendica\LegacyModule\...` namespace

## General recommendations

1. TDD (create tests for new classes/methods or before refactoring a method)
2. Constructor should only be used for dependency injection (called by Dice), but should not have side effects, calling the database, sending an email or anything else.
3. If you want to instantiate a class inside your code use a static factory method instead of `__construct()`
4. Avoid Traits (exceptions are possible)
5. Use abstraction over inheritance (create Interfaces instead of extending classes)
6. Make classes `final` by default
7. Use `private` for properties and methods by default instead of `protected`.

## New Request and Response flow

A typical request and response flow should be work like this:

1. In index.php gather all PHP globals for a request inside a Request object (never call $_GET, $_POST, etc in code inside the `src` folder)
2. Provide the Request object to the App
3. The App gets the `Router` from `Dice::create()`.
4. The Router determines the needed Controller by scanning the Request (HTTP method, uri, headers, etc)
5. The App gets the Controller from `Dice::create()`
6. The Controller is called with the Request and the Controller returns a Response object
8. The Response is returned back to the `index.php`
9. The `index.php` sends the response to the client.

This behavior could be later adapted in the `bin/console.php`

1. In bin/console.php gather all input arguments
2. Provide the arguments to the App (`$app->execute($input, $output);`)

## The new index.php

At the end the `index.php` should look like this:

```php
// Container is created and loaded with the rules
$container = require __DIR__ . '/static/bootstrap.php');

// The App is created by providing the DIC
// Or we move the creation of the App object inside the bootstrap.php too
$app = \Friendica\App::fromContainer($dice);

// Get the ServerRequest that contains every globals like $_GET, $_POST, etc
$request = \GuzzleHttp\Psr7\ServerRequest::fromGlobals();

// Let the app handle the request
$response = $app->handleRequest($request);

// Send the response to the client
// This should be the only place that causes side-effects like `header()`, `echo()` or `exit()`
header(sprintf('HTTP/%s %s %s',
    $response->getProtocolVersion(),
    $response->getStatusCode(),
    $response->getReasonPhrase(),
));

foreach ($response->getHeaders() as $name => $values) {
    header($name . ': ' . $response->getHeaderLine($name));
}

echo $response->getBody();
exit();
```

**Comment by annando** [comment]
How does this whole stuff work with our static classes?

**Comment by Art4** [comment]
This depends on the class. Static classes aren't a problem at all. Only if a static class has dynamic dependencies to other classes (what happens very often) than it should be refactored. The dependency could be added as a new parameter to the method call or be rewritten the class to non-static methods and the dependency is provided to the constructor. It depends on the specific class to find the best refactoring method.

Ideally refactoring should be done in 4 steps:

### 1. create an alternative

If you have a class, method or function that should be refactored create a new class or method with this functionality. Create tests for this new functionality. If there are already tests for the old class then the tests should also work with the new functionality.

### 2. deprecate the old function

If possible make use of the new functionality inside the old functionality. Add a `@deprecated` doctype in the method/class phpdoc section and tell what alternative should be used instead.

### 3. switch to the new functionality

All places that calls the old functionality must be rewritten to use the new functionality. Depending on number of calls this task could be done in one or in multiple PRs.

### 4. remove the old functionality

If there are no more calls on the old and deprecated functionality than it can be deleted. Also the tests for the old functionality can be deleted.


This refactoring cycle can happen in parallel on many places. Depending on the size and number of calls of the old functionality this steps could be happen in one single PR or in many PRs during a longer period of time.

**Comment by nupplaphil** [comment]
just a short history about the `DI` service locator from my point of view.

The intention of using Dice was exactly what you summarize here. I chose it because it's really small and it is fast in case of loading ~100 classes per execution (if we would need to load more than 100 classes per call, there are other frameworks, which are more efficient).

When I introduced it, I came across the same problems, combining the existing "static" classes (which were used for any kind of functionality in the past) and the new paradigm of "loosely coupling". After some trial-and-errors, I started refactoring the `Core` Classes of Friendica (Cache, Lock, Session, Logging, Database ...), because they were mostly testable without any dependencies to other classes.

The `DI` (singleton) was the way to refactor class-by-class without getting a headache because of the high dependencies of classes in Friendica :)

So my way to go was always:
1) Choosing a class
2) Introduce a constructor
3) Try to replace all `DI` calls in the class with dependency loading per instructor
4) Refactor the class from `static` functions to class-methods
5) Maybe split it into  layer-oriented classes (Service, Model, View, ...)
6) Add as many tests as possible
7) Search for every occurrence in Friendica and either add it in the existing constructors or introduce a `DI::` method for it (so I didn't had to refactor other classes as well - which would have been a never ending story ...)

I found it nearly impossible to refactor whole Friendica in using Dependency injection instead of static class initializing, so I tried it step by step but gluing both worlds with `DI` together :)

The Request-Response flow is exactly, what I had in mind as a result of my first refactorings! :-)

**Comment by MrPetovan** [comment]
I'm reading this issue for the first time and I have been delighted that @Art4 reached the same conclusions @nupplaphil laid out when setting up the `DI` helper in the first place. Good job @nupplaphil , you evidently made it straightforward to follow your thought path and pick up the work where we kind of left off!

**Comment by Art4** [comment]
@nupplaphil Thank you very much for your insights. :+1: I'm glad I got most of the decisions I saw in the code architecture right. Regarding the DI class, I didn't get the idea that this should only be a temporary solution. Now I realize that and I think it was the right decision. I would have done it the same way, but would have given the class a name that indicates its temporariness, e.g. `LegacyServiceLocator` or something similar.

The refactoring will probably take years, but if the goal is clear and the path to it is shaped with improvements and deprecations, then it should be possible to modernize Friendica piece by piece. I am very optimistic and look forward to a good collaboration.

**Comment by MrPetovan** [comment]
> I would have done it the same way, but would have given the class a name that indicates its temporariness, e.g. `LegacyServiceLocator` or something similar.

The name of this class has been a hot button issue, with @annando complaining about longer names because of the difficulty to type it manually every time. 😄

**Comment by Art4** [comment]
> > I would have done it the same way, but would have given the class a name that indicates its temporariness, e.g. `LegacyServiceLocator` or something similar.
> 
> The name of this class has been a hot button issue, with @annando complaining about longer names because of the difficulty to type it manually every time. 😄

Long names for temporary (deprecated) classes is also a little tweak in refactoring: make them harder to use so that using them feels wrong and using the newer solution (dependency injection) feels easier. :wink: Also an IDE could help with this.

**Comment by MrPetovan** [comment]
I don’t disagree, but we are working around @annando as he has been the most prolific and consistent contributor of this project over the past decade. As a result any convenience request from his part is carefully considered regardless of its theoretical demerit.

In particular with the introduction of Dependency Injection we’ve pushed him to his limits in terms of abstract understanding, so we’ve made it as convenient as possible for him to keep contributing without feeling overwhelmed. Also it isn’t stubbornness on his part, he’s been using the new paradigm sporadically on his own, we just need to adapt to his pace because he’s irreplaceable for the project at the moment.

**Comment by annando** [comment]
@Art4 I now use an IDE, but nonetheless I would like to have readable code - and too long class names are really hard to read, I think. Also I don't think it would had any influence on transforming the code.

**Comment by Art4** [comment]
@nupplaphil May I ask you to rework the `bin/auth_ejabberd.php` into a Console class as you did with with `bin/console.php` and `bin/worker.php`?

**Comment by Art4** [comment]
I will work on this again in the near future. @annando Can you assign this issue to me so I will find it better? Update: Thank you @nupplaphil 👍 

Some refactoring notes for me after the PHP 8.2 upgrade (#15652):

- [x] #15652
- [x] Bring Rector to `LevelSetList::UP_TO_PHP_85` and `withPhpVersion(80200)`
    - [x] #15656 
    - [x] #15657
    - [x] #15738
    - [x] #15770
    - [x] #15782
    - [x] #15805
    - [x] #15806
- [x] update to PHPUnit 11
    - [x] #15661
    - [x] #15705
    - [x] #15660
- [x] Update or replace all other libraries in `composer.json`
  - [x] #15865
- [x] PHPStan Level 4 
  - [x] #15787
  - [x] #15855
  - [x] #15889
  - [x] #15918
- [x] Fix or remove deprecated code (search for `@deprecated` label and `trigger_error()` calls)
    - [x] Install `phpstan/phpstan-deprecation-rules` and fix all deprecation calls
- [x] #15719
